### PR TITLE
NNCF - PTQ Object Detection & Classification

### PR DESCRIPTION
**Pull request by Akash-guna**
Previous PR #7

### CODE REVIEW

File: quant/__init__.py [hunk 1]
```diff
@@ -1,12 +1,16 @@
-def initialize_initialization(algoname):
+def initialize_initialization(algoname, task):
```
Comment (shwu-nyunai) [inline quant/__init__.py:1]: lets use a more suitable name or simply override the dunder __getattr__

File: core/utils/mmutils.py [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+    metainfo = get_metainfo_coco(data_path)
+    ann_file = os.path.join(data_path, "annotations/instances_val2017.json")
+    cfg = replace_all_instances(
+        cfg, "data_root", data_path, create_additional_parameters={"metainfo": metainfo}
+    )
+    cfg = replace_all_instances(cfg, "ann_file", ann_file)
+    cfg = replace_all_instances(cfg, "batch_size", batch_size)
+    cfg.work_dir = os.path.join(cache_path)
+    runner = Runner.from_cfg(cfg)
+    cfg.dump("current_quant_final.py")
```
Comment (shwu-nyunai) [inline core/utils/mmutils.py:28]: what is the directory where the dumps are happening? also, are these configs cleared after the job is completed? 
if not, lets either delete them post job or lets dump them inside inferred job dirs or cache dirs; 

same for all yamls, jsons, and intermediary files created.

File: quant/nncf/ptq/main.py [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+        self.loaders = loaders
+        self.wandb = kwargs.get("wandb", True)
+        self.imsize = kwargs.get("insize", "32")
+        self.to_train = kwargs.get("TRAINING", True)
+        self.model_name = kwargs.get("MODEL", "resnet50")
+        self.imsize = kwargs.get("insize", "32")
+        self.model_path = kwargs.get("MODEL_PATH", "models")
+        self.logging_path = kwargs.get("LOGGING_PATH", "logs")
+        self.logger = logging.getLogger(__name__)
+        self.logger.info(f"Experiment Arguments: {self.kwargs}")
```
Comment (shwu-nyunai) [inline quant/nncf/ptq/main.py:17]: we don't need to store the loggers in the classes; lets elevate the loggers out of the classes if loggers aren't dependent on the classes. 
same for all such occurrences.

File: quant/nncf/ptq/objectdetection.py [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+            if "model_ptq.pth" in os.listdir(os.path.join(self.model_path, folder)):
+                self.quantized_pth_location = os.path.join(
+                    self.model_path, folder, "model_ptq.pth"
+                )
+                break
+        # deply config
+        build_mmdeploy_config(self.imsize)
+        create_input_image(self.loaders["test"])
+        os.system(
+            f"python vision/core/utils/mmrazordeploy.py current_openvino_deploy_config.py current_quant_config.py {self.quantized_pth_location} demo_image.png"
```
Comment (shwu-nyunai) [inline quant/nncf/ptq/objectdetection.py:74]: lets have at least one runtime inferred path and refer that path for all paths we create. 
ideally, the repo root, or current file, or module root is inferred at runtime as follows and then extended to create the file paths;

```python
from pathlib import Path

ROOT = Path(__file__) # get the root path to module, or repo root 
extended_path = ROOT / "relative" / "path" / "from" / "root"
```

File: quant/nncf/ptq/objectdetection.py [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+        )
+        self.logger.info("Deployment Successful")
+        shutil.move("/workspace/nyuntam/end2end.xml", os.path.join(self.model_path, "mds.xml"))
+        shutil.move("/workspace/nyuntam/end2end.bin", os.path.join(self.model_path, "mds.bin"))
```
Comment (shwu-nyunai) [inline quant/nncf/ptq/objectdetection.py:78]: lets avoid absolute paths; infer these relative to some runtime inferred path.

File: quant/nncf/ptq/utils.py [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+            keep_top_k={keep_top_k},
+            background_label_id=-1,
+        )),
+    function_record_to_pop=[
+        'mmdet.models.detectors.single_stage.SingleStageDetector.forward',
+        'mmdet.models.detectors.two_stage.TwoStageDetector.forward',
+        'mmdet.models.detectors.single_stage_instance_seg.SingleStageInstanceSegmentor.forward'  # noqa: E501
+    ])"""
+    with open("current_base_openvino_deploy_config.py", "w") as f:
+        f.write(cfg)
```
Comment (shwu-nyunai) [inline quant/nncf/ptq/utils.py:58]: nitpick - 
ensure these paths are cleaned after the job completion or are created in cache/job paths

File: quant/nncf/ptq/utils.py [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+def build_mmdeploy_config(insize):
+    config = f"""_base_ = ['vision/core/utils/mmdeployconfigs/mmdet/_base_/base_dynamic.py', 'vision/core/utils/mmdeployconfigs/_base_/backends/openvino.py']
+
+onnx_config = dict(input_shape=None)
+
+backend_config = dict(
+    model_inputs=[dict(opt_shapes=dict(input=[1, 3, {insize}, {insize}]))])
+"""
+    with open("current_openvino_deploy_config.py", "w") as f:
+        f.write(config)
```
Comment (shwu-nyunai) [inline quant/nncf/ptq/utils.py:132]: same as above. ignore if out of concern.

**Review by shwu-nyunai — CHANGES_REQUESTED** [review]

File: quant/nncf/ptq/main.py (hunk 3 above)
Comment (Akash-guna) [reply quant/nncf/ptq/main.py:17]: I suppose this change does not change any workflows / outputs so ignoring this for now

File: quant/__init__.py (hunk 1 above)
Comment (Akash-guna) [reply quant/__init__.py:1]: in a later commit

Comment (shwu-nyunai) [reply quant/__init__.py:1]: okay

File: quant/nncf/ptq/main.py (hunk 3 above)
Comment (shwu-nyunai) [reply quant/nncf/ptq/main.py:17]: okay

**Review by shwu-nyunai — APPROVED** [review]
lgtm

**Review by Arnav0400 — APPROVED** [review]

**Pull request merged**